clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3007,19 +3007,37 @@ class SYCLKernelNameTypeVisitor
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
     const DeclContext *DeclCtx = Tag->getDeclContext();
     if (DeclCtx && !UnnamedLambdaEnabled) {
-      auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
-      if (NameSpace && NameSpace->isStdNamespace()) {
-        S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << KernelNameType;
-        S.Diag(KernelInvocationFuncLoc, diag::note_invalid_type_in_sycl_kernel)
-            << /* kernel name cannot be a type in the std namespace */ 2
-            << QualType(Tag->getTypeForDecl(), 0);
-        IsInvalid = true;
-        return;
+      // Loop through Declaration Contexts to detect kernel names nested inside
+      // "std" and "anonymous" namespaces.
+      while (!DeclCtx->isTranslationUnit()) {
+        auto *NSDecl = dyn_cast<NamespaceDecl>(DeclCtx);
+        if (NSDecl && NSDecl->isStdNamespace()) {
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::err_sycl_kernel_incorrectly_named)
+              << KernelNameType;
+          S.Diag(KernelInvocationFuncLoc,
```

Comment:
I know the original code did this, but why issue a note on the exact same location as the error? Typically, this would be handled in a single diagnostic.

Reply:
The discussion in this [PR](https://github.com/intel/llvm/pull/2730) might give you some background. 
I believe the decision to issue both an error as well as a note came from this [suggestion](https://github.com/intel/llvm/pull/2730#discussion_r517866371)
AFAIR, the explanation given at that point was that "error" would point out what the actual error was, and the "note" would provide the reason for the error.
Example:
error: `'some_kernel_name' is invalid kernel name ` (actual error)
note: `'some_kernel_name' type is not globally-visible` (reason)

Reply:
Hmm, I can understand the logic, but I'm not certain I'd give the same advice. Diagnostic text should be succinct, so I definitely agree with that. But error + note is often *more* text than just issuing a slightly longer error, in two different ways:

1) Amount of actual text shown is longer. e.g.,
```
error: 'some_kernel_name' is invalid kernel name'
note: 'some_kernel_name' type is not globally visible
```
vs
```
error: 'some_kernel_name' is invalid kernel name; type is not globally visible
```
So it's 102 characters for error + note, and 78 characters for just the error.

2) Number of diagnostics displaying in an IDE. Some editors (like MSVC and Xcode) helpfully aggregate diagnostics into a listbox rather than as text on the console and having two diagnostics for the same line of code doubles the number of entries in such a list.

FWIW, in the community side of things, we usually only issue a note at the same location as a related diagnostic when there's a fix-it involved but the fix-it may not always be correct. e.g., "did you mean X?" and "did you mean Y?" kind of fix-its where there are multiple valid ways to resolve the diagnostic. (If the fix-it is unambiguous or never problematic, we issue it directly on the diagnostic rather than via a note.)

Reply:
I can combine the diagnostic text into a single error and drop the note.
Do you have a suggestion for the diagnostic error and do you think this should be discussed with code owners?

Reply:
I'd say we can combine the error and note text (with minor adjustments):
```
def err_sycl_kernel_incorrectly_named : Error<
  "%0 is an invalid kernel %select{name|type}1; "
  "%select{it must be globally visible"
  "|unscoped enum requires a fixed underlying type"
  "|it cannot be in the \"std\" namespace"
  "|unnamed type used in a SYCL kernel name"
  "}2">;
```
(I may not have gotten it exactly right as I'm not familiar with the terminology we use -- things like "unnamed type used in a SYCL kernel name" reads a bit strange to me.)

I think we may want to ensure the code owners are all on the same page regarding diagnostic wordings in general (this note thing has come up before).

Reply:
> (I may not have gotten it exactly right as I'm not familiar with the terminology we use -- things like "unnamed type used in a SYCL kernel name" reads a bit strange to me.)
The wordings for this diagnostic came from Erich. If this needs to be changed, please let me know.
FYI @elizabethandrews @premanandrao
If we all agree on dropping the note and modifying the error as Aaron's suggestion, I will make that change.
I prefer to make this change as part of this PR so I can re-write the code as suggested [here](https://github.com/intel/llvm/pull/3225#discussion_r577915522)

Reply:
Also tagging @erichkeane in case he has opinions on the wording.

Reply:
I agree with Aaron that combining these is a good idea.  

I'm not a huge fan of `it must be globally visible`, but don't have a better suggestion :/  Perhaps remove the 'it' from both the 1st and 3rd one?  What do you think @AaronBallman ?


The "unnamed type used in a kernel name" is a little funky and predates our involvement and is a little silly.  In:

    kernel_call<class Foo>(...);

`Foo` is the 'Kernel Name'.  However, if it is a type we can't name separately (basically, we cannot forward declare it), we have to reject it.  A 'note' for that one should optimally point to the unnamed type, but I don't remember if we did that or not.

Reply:
Ah, thanks for the explanation about what a kernel name is. In that case, I'd go with:
```
def err_sycl_kernel_incorrectly_named : Error<
  "%q0 is an invalid %select{kernel name|type for a kernel name}1; "
  "%select{declaration must be globally visible"
  "|unscoped enum requires a fixed underlying type"
  "|declaration cannot be in the \"std\" namespace"
  "|type cannot be unnamed"
  "}2">;
```
using `kernel name` for 1 and 3, and `type for a kernel name` for 2 and 4. WDYT?

Reply:
I think that works... I typed out the 4 combos and they are a little awkward, but seemingly descriptive enough.  I don't have any better suggestions.

Reply:
I'm ok with combining the error + note but we need to pass types to some of the 'select' options. For example - in the current diagnostic `type %1 cannot be in the std namespace`, %1 does not refer to kernel name. It refers to the type in the kernel name which triggers this diagnostic.
For e.g. 
```
    // expected-error@Inputs/sycl.hpp:220 {{'Templated_kernel_name<nullptr_t>' is an invalid kernel name type}}
    // expected-note@Inputs/sycl.hpp:220 {{type 'nullptr_t' cannot be in the "std" namespace}}
```

It is very important to keep this information. Prior to this change, I spent way too hours trying to find problematic types in templated kernel names.

Reply:
Actually no. I think the issue in the past was that we didn't print fully instantiated name for kernel name. We do that now. So disregard my previous comment. I still think its nice to point to problematic type but I guess it isn't necessary.

Reply:
Based on this https://github.com/intel/llvm/pull/2730#issuecomment-722038199, mentioning problematic type is useful for std types.

Reply:
Oh interesting, then I think my wording is a bit lackluster -- I'd be happy keeping the type information in the diagnostic in that case.

Reply:
@AaronBallman 
Based on the above comments, I am printing the fully instantiated kernel names for all cases and additionally printing the fully-qualified type just for the 'std' ns case.
This is the updated diagnostics:

`def err_invalid_std_type_in_sycl_kernel : Error<"%0 is an invalid kernel name,type %1 cannot be in the \'std\' namespace ">;`
%0 -> fully instantiated kernel name
%1 -> fully-qualified type
```
def err_sycl_kernel_incorrectly_named : Error<
  "%select{%1 should be globally-visible"
  "|unscoped enum %1 requires fixed underlying type"
  "|type cannot be unnamed"
  "}0">;
```
%1 -> fully instantiated kernel-name

Let me know if this is okay.